hugo/static/cgi-bin/extract_citation.cgi

```diff
@@ -0,0 +1,188 @@
+            paper_id = rest[1:].lstrip("0")
+            return (collection_id, rest[0], paper_id if paper_id else "0")
+
+
+def parse_query_string(query_string):
+    """
+    Parse the query string into a dictionary.
+    """
+    return dict(q.split("=") for q in query_string.split("&"))
+
+
+def bib_iterator(f):
+    """
+    An iterator that iterates over bib entries in a file.
+    """
+    entry = ""
+    for line in f:
+        if line.strip() == "}":
+            entry += line
+            yield entry
+            entry = ""
+        else:
+            entry += line
+
+
+def xml_iterator(f):
+    """
+    An iterator for MODS XML entries.
+
+    Format:
+        <?xml version="1.0" encoding="UTF-8"?>
+        <modsCollection xmlns="http://www.loc.gov/mods/v3">
+        <mods ID="acl-2024-long">
+            <titleInfo>
+                <title>Proceedings of the 62nd Annual Meeting of the Association for Computational Linguistics (Volume 1: Long Papers)</title>
+            </titleInfo>
+        ...
+        </modsCollection>
+    """
+    header = ""
+    entry = ""
+    for line in f:
+        if line.startswith("<?xml") or line.startswith("<modsCollection"):
+            header += line
+        if line.strip() == "</mods>":
+            entry += line
+            yield header + entry + "</modsCollection>"
+            entry = ""
+        else:
+            entry += line
+
+
+def endf_iterator(f):
+    """
+    An iterator for EndNote XML entries.
+    """
+    entry = ""
+    for line in f:
+        if line.strip() == "":
+            yield entry
+            entry = ""
+        else:
+            entry += line
+    if entry:
+        yield entry
+
+
+def get_entry(anthology_id, format):
+    """
+    Opens the volumes file corresponding to the format, then grab the
+    associated iterator, and look for an entry matching the anthology_id.
+    """
+    # The iterator is the function {format}_iterator
+    iterator = globals().get(f"{format}_iterator")
+    if iterator:
+        parsed_id = parse_id(anthology_id)
+        volume_id = f"{parsed_id[0]}-{parsed_id[1]}"
+        try:
+            with open(f"../volumes/{volume_id}.{format}") as f:
+                # iterate through the file, reading bibtex entries
+                for entry in iterator(f):
+                    # match either https://preview.aclanthology.org/{branch}/{anthology_id}/
+                    # or https://aclanthology.org/{anthology_id}/ without using regex (good enough)
+                    if "https://" in entry or "aclanthology.org" in entry and f'/{anthology_id}/' in entry:
+                        return entry
+        except FileNotFoundError:
+            pass
+    return None
+
+
+if __name__ == "__main__":
+    # Get the anthology_id from the query string
+    params = parse_query_string(os.environ.get("QUERY_STRING", ""))
+    anthology_id = params.get("anthology_id")
+    format = params.get("format")
+    entry = get_entry(anthology_id, format)
```

Comment:
I am wondering if this needs some kind of input validation. At least on the preview server, I can access the CGI script directly with arbitrary query strings, for example:

https://preview.aclanthology.org/dont-generate-bib-files/cgi-bin/extract_citation.cgi?anthology_id=..%2F..%2F..%2F..%2F..%2Fetc%2Fpasswd

It doesn’t return anything, and I don’t immediately see an exploit here (because of the way we’re constructing filenames and return strings), but might still be a good idea to block such shenanigans early.

Reply:
I looked at this a little, and it seems pretty constrained, but I added a check for valid Anthology IDs.